### DISCUSSION TITLE
For non-page-based formats like epub, how to get the text by chapters?

### DISCUSSION THREAD

**Discussion by workflowsguy** [pymupdf/PyMuPDF#2503] [Looking for help]
I tried to get the text of a epub book by chapter, but could not find any information on how to do that.
PyMuPdf splits chapters into pages according to a rule that is not obvious to me.

What I would like to get in the end is a list of dicts like [{'Chaptername': chaptertext}].

Is this possible?

Thanks!

**Comment by JorjMcKie** [comment]
You probably just have to iterate over the EPUB chapters:
```python
for i in range(doc.chapter_count):
    chapter_page_count = doc.chapter_page_count(i)
    chapter_text = ""
    for j in range(chapter_page_count):
        page = doc[[(i, j)]
        chapter_text += page.get_text()
    print(f"Text of chapter {i}")
    print(chapter_text)
```

**Comment by workflowsguy** [comment]
Thank you for the example, @JorjMcKie . I had overlooked that the `doc` class provides support for chapters.

I tested this with two epub files. Unfortunately, this does not work as expected.
In `PyMuPDF 1.21.1` `chapter_page_count`returns an incorrect value (it seems like for chapter "n", it actually gets the page count for chapter "n+1"...)

**Reply by JorjMcKie** [reply]
If you were right, then the follow test must fail, right?
```python
chapter_sum = 0
for i in range(doc.chapter_count):
    chapter_sum += doc.chapter_page_count(i)
assert chapter_sum == doc.page_count
```

**Reply by JorjMcKie** [reply]
it doesn't in my test cases.

**Comment by workflowsguy** [comment]
I think that from my tests, I have made an incorrect assumption about the relationship between "chapters" and "toc" entries.

I thought I could get the chapter text like in your example and the name of the chapter from the toc. But as I currently understand, those two items do not map to one another (i.e. chapter 1 <> toc entry 1 of level 1)

**Reply by JorjMcKie** [reply]
You are right. Chapter is a fully technical term in the EPUB's file structure. There is no reason to assume / rely on any relationship with the items in the TOC. The latter may be created independently, following other considerations.
